app/client-v2/src/stores/projectStore.ts

```diff
@@ -75,50 +76,76 @@ export const useProjectStore = defineStore("project", {
+        if (activeBatches.size >= maxWorkers) {
+          await Promise.race(activeBatches); // wait for at least 1 batch to finish
+        }
+        const batchPromise = this.computeAmrAndSketch(hashedFileBatch);
+        activeBatches.add(batchPromise);
+
+        batchPromise
+          .catch(() => console.log("error processing batch"))
+          .finally(() => {
+            activeBatches.delete(batchPromise);
+            uploadPercentNumerator++;
+            this.uploadingPercentage = Math.round((uploadPercentNumerator / hashedFileBatches.length) * 100);
+          });
+      }
+    },
+    getOptimalWorkerCount() {
+      return Math.floor(navigator.hardwareConcurrency * 0.5) || 4;
+    },
+    async computeAmrAndSketch(hashedFiles: HashedFile[]): Promise<void> {
+      return new Promise((resolve, reject) => {
         const worker = new Worker("/worker.js");
-        worker.postMessage({ hash: fileHash, fileObject: file });
+        worker.postMessage(hashedFiles);
 
-        worker.onmessage = async (event: MessageEvent<WorkerResponse>) => {
-          this.handleWorkerResponse(file.name, event);
+        worker.onmessage = async (event: MessageEvent<WorkerResponse[]>) => {
+          this.handleWorkerResponse(event.data);
+          worker.terminate();
+          resolve();
         };
+
         worker.onerror = (error) => {
           console.error(error);
-          this.toast.showErrorToast("Ensure uploaded sample file is correct");
+          this.toast.showErrorToast("Ensure uploaded sample files are correct, or try again later.");
+          worker.terminate();
+          reject();
         };
-      }
+      });
     },
-
-    async handleWorkerResponse(filename: string, event: MessageEvent<WorkerResponse>) {
-      const { hash, result, type } = event.data;
-      const parsedAmrOrSketch = JSON.parse(result);
-
-      const matchedHashIndex = this.project.samples.findIndex((sample: ProjectSample) => hash === sample.hash);
-      if (matchedHashIndex !== -1) {
-        this.project.samples[matchedHashIndex][type] = parsedAmrOrSketch;
-      }
-
+    async handleWorkerResponse(samples: WorkerResponse[]) {
+      this.project.samples.push(...samples);
       try {
-        await baseApi.post(
-          `/project/${this.project.id}/${type}/${hash}`,
-          type === WorkerResponseValueTypes.AMR
-            ? parsedAmrOrSketch
-            : {
-                sketch: parsedAmrOrSketch,
-                filename
-              }
-        );
+        await baseApi.post(`/project/${this.project.id}/sample`, samples);
       } catch (error) {
         console.error(error);
-        this.toast.showErrorToast("Ensure uploaded sample file is correct or try again later.");
-        if (matchedHashIndex !== -1) {
-          this.project.samples.splice(matchedHashIndex, 1);
-        }
+        this.toast.showErrorToast("Ensure uploaded sample files is correct or try again later.");
```

Comment:
```suggestion
        this.toast.showErrorToast("Ensure uploaded sample files are correct or try again later.");
```
..but also, we should revisit how useful this is in the light of possibly processing hundreds of samples. Again, it would be better if the user could get some record of which samples failed which part of the process.